### DISCUSSION TITLE
adding libray has Library Manager removing "unused" ESP Async WebServer and espMqttClient

### DISCUSSION THREAD

**Discussion by koerri** [tbnobody/OpenDTU#433] [Q&A] [answered]
Sorry I'm somewhat new to GitHub procedures, so I'm not sure whether this should be filed as a potential bug or if it's just a configuration problem on my side, or some PlatformIO quirk.

I'm looking to connect an OLED display to the OpenDTU ESP32 board, so I added an SSD1306 library via the VScode/PlatformIO Libraries Registry.
But now the build fails because for some reason the Library Manager suddenly considers the ESP Async WebServer and espMqttClient libs to be "unused" and removes them with their dependencies, see log below.
I've tried various libs, always getting the same errors. Also I'm not even including/using the added library.
On Google I've found a suggestion that having explicit Includes for the "unused" libraries in main.cpp might prevent the unwanted removal but that didn't help.

If I remove/uninstall "my" library from the project, on the next build the "unused" libs will be re-installed by the Library Manager and the build does succeed.

I did play around with the [Library Dependency Finder](https://docs.platformio.org/en/stable/librarymanager/ldf.html#id2) settings, but no combination made a difference.

Fresh install of VScode 1.74.0, PlatformIO Core 6.1.5/Home 3.4.3, OpenDTU build 9bfc704 cloned with git version 2.38.2.windows.1 (also deleted the project and cloned again).

Thanks for any help or ideas!

Terminal output:
> Processing generic (board: esp32dev; framework: arduino; platform: espressif32@>=5.2.0)
> .------------------------------------------------------------------------------------------------------------------------------------------------
> 
> Removing unused dependencies...
> Library Manager: Removing ESP Async WebServer @ 1.2.3+sha.1dde9cf
> Library Manager: Removing dependencies...
> Library Manager: Removing AsyncTCP @ 1.1.1
> Library Manager: AsyncTCP@1.1.1 has been removed!
> Library Manager: ESP Async WebServer@1.2.3+sha.1dde9cf has been removed!
> Library Manager: Removing ArduinoJson @ 6.19.4
> Library Manager: ArduinoJson@6.19.4 has been removed!
> Library Manager: Removing RF24 @ 1.4.5
> Library Manager: RF24@1.4.5 has been removed!
> Library Manager: Removing espMqttClient @ 1.3.3+sha.6a3f3a8
> Library Manager: Removing dependencies...
> Library Manager: espMqttClient@1.3.3+sha.6a3f3a8 has been removed!
> Library Manager: Installing nrf24/RF24 @ >=1.4.2
> Unpacking  [####################################]  100%
> Library Manager: RF24@1.4.5 has been installed!
> Verbose mode can be enabled via `-v, --verbose` option
> Firmware Revision: g9bfc704
> CONFIGURATION: https://docs.platformio.org/page/boards/espressif32/esp32dev.html
> PLATFORM: Espressif 32 (5.2.0) > Espressif ESP32 Dev Module
> HARDWARE: ESP32 240MHz, 320KB RAM, 4MB Flash
> DEBUG: Current (cmsis-dap) External (cmsis-dap, esp-bridge, esp-prog, iot-bus-jtag, jlink, minimodule, olimex-arm-usb-ocd, olimex-arm-usb-ocd-h, olimex-arm-usb-tiny-h, olimex-jtag-tiny, tumpa)
> PACKAGES:
>  - framework-arduinoespressif32 @ 3.20005.220925 (2.0.5)
>  - tool-esptoolpy @ 1.40201.0 (4.2.1)
>  - tool-openocd-esp32 @ 2.1100.20220706 (11.0)
>  - toolchain-xtensa-esp32 @ 8.4.0+2021r2-patch3
> LDF: Library Dependency Finder -> https://bit.ly/configure-pio-ldf
> LDF Modes: Finder ~ chain, Compatibility ~ soft
> Found 38 compatible libraries
> Scanning dependencies...
> Dependency Graph
> |-- U8g2 @ 2.34.6
> |   |-- SPI @ 2.0.0
> |-- LittleFS @ 2.0.0
> |   |-- FS @ 2.0.0
> |-- Hoymiles @ 0.0.1
> |   |-- RF24 @ 1.4.5
> |   |   |-- SPI @ 2.0.0
> |   |-- Every
> |   |-- SPI @ 2.0.0
> |   |-- TimeoutHelper
> |-- Ticker @ 2.0.0
> |-- Ethernet @ 2.0.0
> |   |-- WiFi @ 2.0.0
> |-- WiFi @ 2.0.0
> |-- Update @ 2.0.0
> |-- ResetReason
> |-- DNSServer @ 2.0.0
> |   |-- WiFi @ 2.0.0
> Building in release mode
> Converting .pio\build\generic\index.html.gz.txt.o
> Converting .pio\build\generic\zones.json.gz.txt.o
> Converting .pio\build\generic\favicon.ico.txt.o
> Converting .pio\build\generic\app.js.gz.txt.o
> Compiling .pio\build\generic\src\Configuration.cpp.o
> Compiling .pio\build\generic\src\MqttHassPublishing.cpp.o
> Compiling .pio\build\generic\src\MqttPublishing.cpp.o
> Compiling .pio\build\generic\src\MqttSettings.cpp.o
> Compiling .pio\build\generic\src\NetworkSettings.cpp.o
> Compiling .pio\build\generic\src\NtpSettings.cpp.o
> Compiling .pio\build\generic\src\Utils.cpp.o
> Compiling .pio\build\generic\src\WebApi.cpp.o
> Compiling .pio\build\generic\src\WebApi_config.cpp.o
> Compiling .pio\build\generic\src\WebApi_devinfo.cpp.o
> Compiling .pio\build\generic\src\WebApi_dtu.cpp.o
> In file included from include/WebApi.h:4,
>                  from src/WebApi.cpp:5:
> include/WebApi_config.h:4:10: fatal error: ESPAsyncWebServer.h: No such file or directory
> 
> ***************************************************************************
> * Looking for Compiling .pio\build\generic\src\WebApi_eventlog.cpp.o
> ESPAsyncWebServer.h dependency? Check our library registry!
> *
> * CLI  > platformio lib search "header:ESPAsyncWebServer.h"
> * Web  > https://registry.platformio.org/search?q=header:ESPAsyncWebServer.h
> *
> ***************************************************************************
> 
>  #include <ESPAsyncWebServer.h>
>           ^~~~~~~~~~~~~~~~~Compiling .pio\build\generic\src\WebApi_firmware.cpp.o ~~~~
> compilation terminated.
> Compiling .pio\build\generic\src\WebApi_inverter.cpp.o
> In file included from src/WebApi_config.cpp:5:
> include/WebApi_config.h:4:10: fatal error: ESPAsyncWebServer.h: No such file or directory
> 
**[multiple similar error messages omitted]**
> 
> *********************************************************************
> 
>  #include <ArduinoJson.h>
>           ^~~~~~~~~~~~~~~
> compilation terminated.
> *** [.pio\build\generic\src\Configuration.cpp.o] Error 1
> In file included from src/MqttHassPublishing.cpp:5:
> include/MqttHassPublishing.h:6:10: fatal error: ArduinoJson.h: No such file or directory
> 
> *********************************************************************
> * Looking for ArduinoJson.h dependency? Check our library registry!
> *
> * CLI  > platformio lib search "header:ArduinoJson.h"
> * Web  > https://registry.platformio.org/search?q=header:ArduinoJson.h
> *
> *********************************************************************
> 
>  #include <ArduinoJson.h>
>           ^~~~~~~~~~~~~~~
> compilation terminated.
> *** [.pio\build\generic\src\MqttHassPublishing.cpp.o] Error 1
> In file included from src/MqttPublishing.cpp:6:
> include/MqttSettings.h:7:10: fatal error: espMqttClient.h: No such file or directory
> 
> ***********************************************************************
> * Looking for espMqttClient.h dependency? Check our library registry!
> *
> * CLI  > platformio lib search "header:espMqttClient.h"
> * Web  > https://registry.platformio.org/search?q=header:espMqttClient.h
> *
> ***********************************************************************
> 
>  #include <espMqttClient.h>
>           ^~~~~~~~~~~~~~~~~
> compilation terminated.
> *** [.pio\build\generic\src\MqttPublishing.cpp.o] Error 1
> In file included from src/MqttSettings.cpp:5:
> include/MqttSettings.h:7:10: fatal error: espMqttClient.h: No such file or directory
> 
> ***********************************************************************
> * Looking for espMqttClient.h dependency? Check our library registry!
> *
> * CLI  > platformio lib search "header:espMqttClient.h"
> * Web  > https://registry.platformio.org/search?q=header:espMqttClient.h
> *
> ***********************************************************************
> 
>  #include <espMqttClient.h>
>           ^~~~~~~~~~~~~~~~~
> compilation terminated.
> *** [.pio\build\generic\src\MqttSettings.cpp.o] Error 1
> ================================================================================================ [FAILED] Took 6.72 seconds ================================================================================================
> 
> Environment    Status    Duration
> -------------  --------  ------------
> generic        FAILED    00:00:06.723
> ========================================================================================== 1 failed, 0 succeeded in 00:00:06.723 ==========================================================================================

**Comment by koerri** [comment]
For verification I just installed VScode/PIO on a Linux machine, git-cloned https://github.com/tbnobody/OpenDTU.git, did a successful build, added FastLED via the PIO Home -> Libraries -> Registry tab and on the next build had the same errors as described above.

If someone else could give it a try and check if they see the same behaviour on their setup, that would be really helpful. 
Thanks!

**Comment by tbnobody** [comment, marked as answer]
dont add the library using the UI... just add it to the platformio.ini as every other library...

**Reply by koerri** [reply]
You make it sound so easy! This being the first library to be added to my first PIO project to my first microcontroller project, it's all a bit confusing. :)
But manually adding the library did work, thanks!

Is this a problem with PIO in general or are there special circumstances with the OpenDTU build configuration?

**Reply by tbnobody** [reply]
This should be the case for all more advanced projects which are using the platformio.ini for locally installed libraries (e.g. Wled)
